enhancements/network/nodeport-range-configuration.md

```diff
@@ -0,0 +1,125 @@
+than that allowed by the default service node-port range, it is important to allow dynamically changing this
+range to allow for more ports. This enhancement focuses on the mechanics of allowing users to configure a
+custom service node-port range for their clusters.
+
+### Goals
+
+- Customers can configure the service node-port range for their cluster during install.
+- Customers can configure the service node-port range for their cluster post install as their needs change.
+
+### Non-Goals
+
+- Services launched prior to customers changing the service node-port range will not work if the new range is
+  not inclusive of the old range.
+- This enhancement makes no guarantees of providing non-disruptive change to the service node-port range.
+  In other words customers, when they make non-inclusive changes to the service node-port range are responsible
+  for deleting old services and recreating new services that will then get node ports assigned from the
+  new range.
+- Configuration of multiple service node-port ranges even though supported by the Kubernetes API server will not be
+  allowed. Users are responsible for configuring the range to be large enough for their needs.
+  
+## Proposal
+
+The config observation facility in cluster-kube-apiserver-operator will be used to observe the openshift/api's network 
+configuration object for changes to the service node port range. The config observation controller will then apply the
+change to the `apiServerArguments.service-node-port-range` field. There will only be one entry allowed in the slice.
+
+The following changes are proposed and implemented:
+
+- API changes: [PR #660](https://github.com/openshift/api/pull/660)
+- Library-go changes: [PR #826](https://github.com/openshift/library-go/pull/826)
+- cluster-kube-apiserver-operator changes: [PR #894](https://github.com/openshift/cluster-kube-apiserver-operator/pull/894)
+
+Additionally, to realize the configuration of service node port range, the necessary ports need to be opened on the 
+individual nodes. Traditionally, this has been the responsibility of the Openshift Installer. But since we want this
+configuration to be changed post-install, we need some mechanism to allow for a new set of ports to be opened by _an_
+_entity_ that knows how to tweak the infrastructure.
+
+### Implementation Details/Notes/Constraints
+
+#### Early CI
+
+#### Upstream Kubernetes
+
+This feature is already supported in upstream Kubernetes.
+
+#### Installer
+
+The installer has been traditionally responsible to [open the ports](https://github.com/openshift/installer/blob/master/data/data/aws/vpc/sg-worker.tf#L146)
+belonging to the service node-port range on individual platforms. That functionality will need to be handled via some
+other entity that can also monitor for changes to the service node-port range and open up the necessary ports while the cluster 
+is up and running.
```

Comment:
https://github.com/openshift/installer/issues/1218 has some discussion, and there's also https://issues.redhat.com/browse/SDN-442. It looks like we'd reached consensus that at some point CNO would be responsible for setting up the rules opening up the VXLAN/Geneve ports, but not anything else. So this would be another thing on top of that.

Reply:
I am looking at the machine api operator and currently thinking that is where the opening of ports should belong. but the lines are blurred because the ports and such depends on part of the config that comes from `cluster` network API object. Does it make sense for the CNO to reconcile the network API object and create a CR instance for the machine api operator to consume?

Reply:
The machine api operator appears to not be hostnetwork (https://github.com/openshift/machine-api-operator/blob/master/install/0000_30_machine-api-operator_11_deployment.yaml) so it can't come up until after the VXLAN/Geneve ports are opened.

And I'm not sure machine-api-operator the right place for this anyway; MAO seems very focused on individual machines/nodes/hosts, whereas the problem we're trying to solve here is either with the cluster as a whole, or with the network specifically, not with individual machines.

Particularly with the "needs to be hostNetwork" problem, it may be that there is no better place to do this than in CNO itself.

Reply:
The only pressing reason I mentioned MAO is because it already has the infrastructure API abstractions for nodes. If that is all abstracted out as a library that we can re-use in CNO, then that makes sense. Also, I think was some recent news about a node operator to set all the kernel sysctl's. That may be useful for setting `ip_local_reserved_ports` and `ip_local_port_range`.